File: lc/DynamicProgramming/DP13DecodeWays.java
package lc.DynamicProgramming;

import java.util.*;

/**
 * https://leetcode.com/problems/decode-ways/
 */
public class DP13DecodeWays {
    public static void main(String[] args) {
        DPH02Solution sol = new DPH02SolutionTwo();
        assert sol.numDecodings("12") == 2;
        assert sol.numDecodings("226") == 3;
        assert sol.numDecodings("0") == 0;
        assert sol.numDecodings("20111") == 3;
        assert sol.numDecodings("11106") == 2;
        assert sol.numDecodings("1123") == 5;
        assert sol.numDecodings("06") == 0;
    }
}

// https://leetcode.com/problems/decode-ways/solutions/4454037/97-43-easy-solution-with-explanation/
class DPH02SolutionThree implements DPH02Solution {
    public int numDecodings(String s) {
        int length = s.length();
        if(s.charAt(0) == '0') {
            return 0;
        }
        int[] dp = new int[length + 1];
        dp[0] = 1;
        dp[1] = 1;
        for(int i=2;i<=length;i++) {
            int oneDigit = Integer.parseInt(s.substring(i-1, i));
            int twoDigit = Integer.parseInt(s.substring(i-2, i));
            if(oneDigit != 0) {
                dp[i]+=dp[i-1];
            }
            if(twoDigit>=10 && twoDigit <=26) {
                dp[i]+=dp[i-2];
            }
        }
        return dp[length];
    }
}


interface DPH02Solution {
    public int numDecodings(String s);
}

class DPH02SolutionTwo implements DPH02Solution {
    // TLE
    private boolean isValidDecoding(String s) {
        if(s.length() == 0 || s.charAt(0) == '0') {
            return false;
        }
        int value = Integer.parseInt(s);
        if(value >=1 && value <= 26) {
            return true;
        }
        return false;
    } 
    public int numDecodings(String s) {
        List<String> nodes = new ArrayList<>();
        String firstChar = s.charAt(0) + "";
        if(isValidDecoding(firstChar)) {
            nodes.add(firstChar);
        }
        for(int i=1;i<s.length();i++) {
            List<String> newNodes = new ArrayList<>();
            String ch = s.charAt(i) + "";
            boolean isChValid = isValidDecoding(ch);
            for(String ls: nodes) {
                if(isChValid) {
                    newNodes.add(ch);
                }
                String ns = ls + ch;
                if(isValidDecoding(ns)) {
                    newNodes.add(ns);
                }
            }
            nodes = newNodes;
        }
        return nodes.size();
    }
}

File: lc/DynamicProgramming/DP13DecodeWays.mjs
import assert from "assert";
// https://leetcode.com/problems/decode-ways/description/
// dynamic-programming
/**
 * @param {string} s
 * @return {number}
 */
var numDecodings = function(s) {
    const n = s.length;
    const chars = s.split("");
    const dp = Array.from({ length: n + 1 }).fill(0);
    const util = (index) => {
        if(index<0) {
            return 1;
        }
        if(dp[index+1] === 0) {
            let ways = 0;
            const n0 = parseInt(chars[index], 10);
            if(n0 !== 0) {
                ways += util(index-1);
            }
            if(index >=1 && chars[index-1] !== '0') {
                const n1 = parseInt(chars.slice(index-1, index+1).join(""), 10);
                if(n1 >=10 && n1 <= 26) {
                    ways += util(index-2);
                }
            }
            dp[index+1] = ways;
        }
        return dp[index+1];
    }
    const res = util(n-1);
    return res;
};

assert(numDecodings("12") === 2);
assert(numDecodings("226") === 3);
assert(numDecodings("11106") === 2);
